gplugins/lumerical/eme.py

```diff
@@ -0,0 +1,1063 @@
+                        algo = not algo
+
+                    if algo:
+                        ss.mesh_cells_per_wavelength += 1
+                    else:
+                        for i in range(1, len(ss.group_cells) - 1):
+                            ss.group_cells[i] += 1
+                else:
+                    ss.mesh_cells_per_wavelength += 1
+                    for i in range(1, len(ss.group_cells) - 1):
+                        ss.group_cells[i] += 1
+
+                # Print convergence step
+                if plot:
+                    logger.info(
+                        f"Mesh cells per wavelength: {ss.mesh_cells_per_wavelength} | Num cells: {ss.group_cells} | S-Param Diff: {sparam_diff}"
+                    )
+
+                # Check for convergence
+                if sparam_diff < cs.sparam_diff:
+                    converged = True
+                else:
+                    converged = False
+            else:
+                # Update mesh and number of cells
+                ss.mesh_cells_per_wavelength += 1
+                for i in range(1, len(ss.group_cells) - 1):
+                    ss.group_cells[i] += 1
+
+        if plot:
+            plt.figure()
+            plt.plot(mesh_cells_per_wavl, s21)
+            plt.plot(mesh_cells_per_wavl, s11)
+            plt.xticks(
+                mesh_cells_per_wavl,
+                [
+                    f"Mesh Cells: {mesh_cells_per_wavl[i]} | Num Cells: {num_cells[i]}"
+                    for i in range(0, len(mesh_cells_per_wavl))
+                ],
+            )
+            plt.setp(plt.xticks()[1], rotation=75, horizontalalignment="center")
+            plt.legend(["|S21|^2", "|S11|^2"])
+            plt.grid("on")
+            plt.xlabel("Mesh Cells Per Wavelength | Num Cells")
+            plt.ylabel("Magnitude")
+            plt.title(f"Overall Convergence | Wavelength={ss.wavelength}um")
+            plt.tight_layout()
+            plt.savefig(
+                str(
+                    self.dirpath
+                    / f"{self.component.name}_convergence"
+                    / f"{self.component.name}_eme_overall_convergence.png"
+                )
+            )
+
+        convergence_data = pd.DataFrame.from_dict(
+            {
+                "mesh_cells": mesh_cells_per_wavl,
+                "num_cells": num_cells,
+                "s21": list(s21),
+                "s11": list(s11),
+            }
+        )
+        convergence_data.to_csv(
+            str(
+                self.dirpath
+                / f"{self.component.name}_convergence"
+                / f"{self.component.name}_eme_overall_convergence.csv"
+            )
+        )
+        return convergence_data
+
+    def get_length_sweep(
+        self,
+        start_length: float = 1,
+        stop_length: float = 100,
+        num_pts: int = 100,
+        group: int = 1,
+    ) -> pd.DataFrame:
+        """
+        Get length sweep sparams.
```

Comment:
what is a length sweep? is this specific to the taper example or is it generically applicable to EME components?

Reply:
this is a sweep of the simulation group span, right? might want to specify that we aren't actually sweeping the length of the device (unless I'm mistaken and we are, in which case clarify that instead)